### ISSUE TITLE
New schema syntax discussion

### ISSUE THREAD

**Issue by guyskk** [guyskk/validr#15]
### Goals

- Easy to write schema in python/high-level programming language, fewer mistakes
- Easy to share and refer schema, and define openapi
- Compatible with YAML, replace `&` with `.` and remove `@` grammar

Overview:

```
# old
name?str&strip&default="world"&desc="Your name"
# new-yaml
name: str.strip.default="world".desc="Your name"
# new-python
name: T.str.strip.default("world").desc("your name")
# new refer
pet: T.ref("http://example.com/schema.json#Pet").optional.desc('description')
```

### Syntax in YAML

```yaml
scalar:
    validator.bool.key=value
list:
    - validator.bool.key=value
    - arg0
    - arg1
dict:
    $self: validator.bool.key=value
    key0: value
    key1: value
refer:
    pet: ref("http://example.com/schema.json#Pet").optional.desc('description')
    pet:
        - ref.optional.desc('description')
        - http://example.com/schema.json#Pet
```

### Syntax in Python

```python
from validr import T

Welcome = T.dict(
    message='str.desc="Welcome message"'
    message=T.str.desc("Welcome message")
).optional.desc('Welcome Object')

@route('/')
def welcome(
    name: 'str.strip.default="world".desc="Your name"',
    name: T.str.strip.default("world").desc("Your name"),
) -> T.list(Welcome).minlen(3):
    return [{'message': 'hello ' + name}] * 3
```

**Comment by guyskk** [comment]
See https://github.com/guyskk/validr/tree/next

```
pip install cython
VALIDR_USE_CYTHON=1 pip install git+https://github.com/guyskk/validr.git@next#egg=validr
```

**Comment by guyskk** [comment]
Another syntax(not implemented yet):
```python
T[ XXX ]       # equal to T.list( XXX )
T(key=XXX)     # equal to T.dict(key=XXX)
```
Is it better? Which one do you prefer?

**Comment by guyskk** [comment]
[new features implemented] modelclass and schema slice:

```python
from validr import T, modelclass, asdict

@modelclass
class Model:
    """Base Model"""

class Person(Model):
    name=T.str.maxlen(16).desc('at most 16 chars')
    website=T.url.optional.desc('website is optional')

guyskk = Person(name='guyskk', website='https://github.com/guyskk')
print(asdict(guyskk))

# schema slice
slice1 = Person['name', 'website']
slice2 = T(Person)['name', 'website']
assert slice1 == slice2
```

**Comment by guyskk** [comment]
closed by #16
